experimental/rocm/rocm_driver.c

```diff
@@ -0,0 +1,190 @@
+                                                                  iree_host_size_t *out_device_info_count) {
+  iree_hal_rocm_driver_t *driver = iree_hal_rocm_driver_cast(base_driver);
+  // Query the number of available ROCM devices.
+  int device_count = 0;
+  ROCM_RETURN_IF_ERROR(&driver->syms, hipGetDeviceCount(&device_count), "hipGetDeviceCount");
+
+  // Allocate the return infos and populate with the devices.
+  iree_hal_device_info_t *device_infos = NULL;
+  iree_host_size_t total_size = device_count * sizeof(iree_hal_device_info_t);
+  for (iree_host_size_t i = 0; i < device_count; ++i) {
+    total_size += IREE_MAX_ROCM_DEVICE_NAME_LENGTH * sizeof(char);
+  }
+  iree_status_t status = iree_allocator_malloc(host_allocator, total_size, (void **)&device_infos);
+  if (iree_status_is_ok(status)) {
+    uint8_t *buffer_ptr = (uint8_t *)device_infos + device_count * sizeof(iree_hal_device_info_t);
+    for (iree_host_size_t i = 0; i < device_count; ++i) {
+      hipDevice_t device;
+      iree_status_t status = ROCM_RESULT_TO_STATUS(&driver->syms, hipDeviceGet(&device, i), "hipDeviceGet");
+      if (!iree_status_is_ok(status)) break;
+      buffer_ptr = iree_hal_rocm_populate_device_info(device, &driver->syms, buffer_ptr, &device_infos[i]);
+    }
+  }
+  if (iree_status_is_ok(status)) {
+    *out_device_info_count = device_count;
+    *out_device_infos = device_infos;
+  } else {
+    iree_allocator_free(host_allocator, device_infos);
+  }
+  return status;
+}
+
+static iree_status_t iree_hal_rocm_driver_select_default_device(iree_hal_rocm_dynamic_symbols_t *syms,
+                                                                int default_device_index,
+                                                                iree_allocator_t host_allocator,
+                                                                hipDevice_t *out_device) {
+  int device_count = 0;
+  ROCM_RETURN_IF_ERROR(syms, hipGetDeviceCount(&device_count), "hipGetDeviceCount");
+  iree_status_t status = iree_ok_status();
+  if (device_count == 0 || default_device_index >= device_count) {
+    status = iree_make_status(IREE_STATUS_NOT_FOUND, "default device %d not found (of %d enumerated)",
+                              default_device_index, device_count);
+  } else {
+    hipDevice_t device;
+    ROCM_RETURN_IF_ERROR(syms, hipDeviceGet(&device, default_device_index), "hipDeviceGet");
+    *out_device = device;
+  }
+  return status;
+}
+
+static iree_status_t iree_hal_rocm_driver_create_device(iree_hal_driver_t *base_driver, iree_hal_device_id_t device_id,
+                                                        iree_allocator_t host_allocator,
+                                                        iree_hal_device_t **out_device) {
+  iree_hal_rocm_driver_t *driver = iree_hal_rocm_driver_cast(base_driver);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // IREE_RETURN_AND_END_ZONE_IF_ERROR(z0, ROCM_RESULT_TO_STATUS(&driver->syms, hipInit(0), "hipInit"));
```

Comment:
You don't need hipInit()? I would remove the commented out line.

Reply:
Forgot to uncomment, was testing something :)